### PR TITLE
fix BptreeMap doube free on remove

### PR DESCRIPTION
**Pull request by Erigara**
I've got double free on attempt to remove element from `BptreeMap`.

Here example i was able to find:

```rust
#[test]
fn bptree_remove_1() {
    let values = [
        4u8, 9, 12, 27, 34, 40, 59, 81, 89, 100, 142, 183, 189, 196, 218, 241,
    ];

    let to_remove = [9u8, 27, 40, 4].map(|v| v.to_string());

    let bptree_map = BptreeMap::from_iter(
        values
            .iter()
            .cloned()
            .map(|v| (v.to_string(), v.to_string())),
    );
    let mut bptree_map_write_tx = bptree_map.write();

    for key in to_remove {
        assert!(bptree_map_write_tx.remove(&key).is_some());
    }
}
```

I've manged to debug that `start_idx` inside `take_from_l_to_r` was not correct when removing keys.
Though there is might be other corner cases not yet found. 

Also i've added proptests to check consistency of `BptreeMap` with `BTreeMap` from std.

### CODE REVIEW

File: tests/bptree_map.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+        let btree_value = btree_map.get(&key);
+        let bptree_value = bptree_map_read_tx.get(&key);
+
+        assert_eq!(btree_value, bptree_value);
+    }
+
+    #[test]
+    fn bptree_remove_consistent(values in proptest::collection::btree_set(proptest::arbitrary::any::<u8>(), 1..256), indices: Vec<proptest::sample::Index> ) {
+        let mut btree_map = BTreeMap::from_iter(values.iter().cloned().map(|v| (v.to_string(), v.to_string())));
```
Comment (yaleman) [inline tests/bptree_map.rs:36]: Minor nitpick, can you remove all the `to_string`'s here and in the next test - it's an unnecessary call and I couldn't work out why they're there.

**Review by yaleman — CHANGES_REQUESTED** [review]

File: tests/bptree_map.rs (hunk 1 above)
Comment (Erigara) [reply tests/bptree_map.rs:36]: `to_strings` were necessary there to spot the issue, because strings require allocation and so we can spot double free

Comment (yaleman) [reply tests/bptree_map.rs:36]: `bptree_map::bptree_remove_consistent` doesn't fail before the patch in `src/internals/bptree/node.rs`...  `bptree_remove_1` does however.

Comment (Erigara) [reply tests/bptree_map.rs:36]: i've just tried to create `Box<u8>` but this way i can't reproduce the issue :(

Comment (Erigara) [reply tests/bptree_map.rs:36]: > bptree_map::bptree_remove_consistent doesn't fail before or after the patch 

interesting, on my machine `bptree_remove_consistent` fails every single time with errors like: 

```
lib-1f4d753d0d073d17(93725,0x16f473000) malloc: Double free of object 0x137708cd0
lib-1f4d753d0d073d17(93725,0x16f473000) malloc: *** set a breakpoint in malloc_error_break to debug
```

```
lib-1f4d753d0d073d17(93197,0x16f987000) malloc: Corruption of free object 0x1447079c0: msizes 37632/0 disagree
lib-1f4d753d0d073d17(93197,0x16f987000) malloc: *** set a breakpoint in malloc_error_break to debug
```

Comment (yaleman) [reply tests/bptree_map.rs:36]: Curious, what machine/os etc?

Comment (yaleman) [reply tests/bptree_map.rs:36]: Actually it fails now on rust v1.79, so I'm happy with this as-is.

**Review by yaleman — APPROVED** [review]
